### PR TITLE
Refactor ServerSideRender to use React hooks.

### PR DESCRIPTION
**Pull request by ZebulanStanphill**
## Description
Part of #22890. This PR refactors the `ServerSideRender` component into a function component using React hooks.

Getting this merged will make future changes like #28289 easier to land.

## How has this been tested?
I inserted an Archives block (which uses `ServerSideRender`) and verified that changing the options in the toolbar and inspector updated the rendered block in the editor.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: packages/server-side-render/src/server-side-render.js
Comment (fabiankaegy) [inline packages/server-side-render/src/server-side-render.js:66]: I would refrain from using the function name `fetch` because it is already available on the global window object.

**Review by fabiankaegy — CHANGES_REQUESTED** [review]

File: packages/server-side-render/src/server-side-render.js
Comment (ZebulanStanphill) [reply packages/server-side-render/src/server-side-render.js]: Good point. Addressed. :+1:

File: packages/server-side-render/src/server-side-render.js
Comment (kevin940726) [inline packages/server-side-render/src/server-side-render.js:140]: Nitpicking: I think we can potentially tidy these up into just one effect. IMHO it's clearer and we can save the need for comparing `isEqual` on every render.

```js
const isMountedRef = useRef(false);

useEffect(() => {
  function fetchData() {
    // ...
  }

  // Initial request
  if (!isMountedRef.current) {
    fetchData();
  }

  isMountedRef.current = true;

  const debounceFetchData = debounce(fetchData, 500);
  
  // Debounced request
  debounceFetchData();

  return () => {
    isMountedRef.current = false;
  }
}, [...])
```

It's not tested so I might have missed something though.

Comment (ZebulanStanphill) [reply packages/server-side-render/src/server-side-render.js:140]: I don't think the suggested code would work:

- It would cause `fetchData` to be called twice on the first render.
- It would cause `debounceFetchData` to be called on every render, which would cause `response` to be set on every render, thus causing an infinite loop, if I had to guess. The comparison against the previous props exists to prevent this.

Comment (kevin940726) [reply packages/server-side-render/src/server-side-render.js:140]: Ahh, yeah, you're right, I missed that part. I still believe we can somehow skip the `isEqual` comparison though, but that's not very important anyway.

Comment (ZebulanStanphill) [reply packages/server-side-render/src/server-side-render.js:140]: There's a pretty good article I found recently discussing this kinda stuff:

https://www.benmvp.com/blog/object-array-dependencies-react-useEffect-hook/

In this case, the `useEffect` callback is dependent upon `attributes`, `block`, `urlQueryArgs`, and `httpMethod`; the first 3 of those are objects (meaning we can't just use shallow equality checks), and using destructuring on the first 2 won't work since their properties may themselves be objects.

As suggested in the article, a `useDeepCompareEffect` hook would be the best way to solve the problem, but that would probably only be a slight improvement over what I'm already doing, and since it introduces an external dependency (albeit a small one), I think it might be better suited to a separate PR.

Comment (kevin940726) [reply packages/server-side-render/src/server-side-render.js:140]: Yeah, I think this is good enough, and we're just refactoring it to hooks anyway. 👍

**Review by kevin940726 — APPROVED** [review]
LGTM 👍 , not tested though. I guess it should be fine as long as the tests are passing?

**Comment by ZebulanStanphill** [comment]
I went ahead and did a quick test again and it seems like everything is working as intended.

@fabiankaegy You need to dismiss your review or switch to an approving review before I can merge. For some reason I can't find a way to do the former. Once this is merged, #28289 can be rebased and will be a lot smaller and easier to review, which should hopefully help it to land in core sooner.

**Review by fabiankaegy — APPROVED** [review]

**Comment by pryley** [comment]
@fabiankaegy The problem with this change is that if you change a prop (block option) during the loading stage (while the spinner is showing), the entire content of the block disappears.

## Before

https://user-images.githubusercontent.com/134939/135711766-d85154f5-c9a2-4447-aabf-70fadc8235b4.mp4

## After

https://user-images.githubusercontent.com/134939/135711774-c450a095-bacb-44ea-91b5-9f37dbf609fd.mp4

FYI: You'll notice in the "after" screen recording that the layout of the block changes during loading. This is because I have applied the `useEffect` feature request suggestion here: https://github.com/WordPress/gutenberg/issues/35294 to trigger some custom javascript in the block after render is complete to init a carousel. However, this has nothing to do with the problem described above.

**Pull request merged**
